mlir/lib/Conversion/GPUToLLVMSPV/GPUToLLVMSPV.cpp

```diff
@@ -259,23 +258,33 @@ struct GPUShuffleConversion final : ConvertOpToLLVMPattern<gpu::ShuffleOp> {
   }
 
   /// Get the subgroup size from the target or return a default.
-  static int getSubgroupSize(Operation *op) {
-    return spirv::lookupTargetEnvOrDefault(op)
-        .getResourceLimits()
-        .getSubgroupSize();
+  static std::optional<int> getSubgroupSize(Operation *op) {
+    auto module = op->getParentOfType<gpu::GPUModuleOp>();
+    if (!module)
+      return std::nullopt;
+    auto spatialExtents = module.getSpatialExtents();
+    if (!spatialExtents)
+      return std::nullopt;
+    return spatialExtents->getReqdSubgroupSize();
   }
 
-  static bool hasValidWidth(gpu::ShuffleOp op) {
+  static bool hasValidWidth(gpu::ShuffleOp op, unsigned subgroupSize) {
     llvm::APInt val;
     Value width = op.getWidth();
-    return matchPattern(width, m_ConstantInt(&val)) &&
-           val == getSubgroupSize(op);
+    return matchPattern(width, m_ConstantInt(&val)) && val == subgroupSize;
   }
 
   LogicalResult
   matchAndRewrite(gpu::ShuffleOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const final {
-    if (!hasValidWidth(op))
+    auto subgroupSize = getSubgroupSize(op);
+    if (!subgroupSize) {
+      return rewriter.notifyMatchFailure(
+          op, "gpu.module spatial_extents attr with reqdSubgroupSize needed "
+              "for conversion");
```

Comment:
```suggestion
          op, "spatialExtents attribute with reqdSubgroupSize in parent symbol table needed "
              "for conversion");
```
This is the correct name, right?

Reply:
`spatial_extents` is the mnemonic, I thought that would be more appropriate, but I'm happy to change it.

Reply:
Up to you!